File: ruby/src/detectors/path-traversal/path-traversal.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=path-traversal@v1.0 defects=1}
def render_modern_param_noncompliant
    page = params[:page]
    # Noncompliant: Unsanitized user-input is used in render file.
    render file: "/some/path/#{page}"
end
# {/fact}

# {fact rule=path-traversal@v1.0 defects=0}
def render_modern_param_compliant
    page = params[:page]
    # Compliant: User-input is sanitized before using it in render file.
    render file: File.basename("/some/path/#{page}")
end
# {/fact}

File: csharp/src/detectors/path-traversal/PathTraversal.cs
namespace Traversal
{
    class PathTraversal
    {
        // {fact rule=path-traversal@v1.0 defects=1}
        public static byte[] PathTraversalNoncompliant(string filename)
        {
            if (string.IsNullOrEmpty(filename)) 
            {   
                throw new ArgumentNullException("error"); 
            }
            string filepath = Path.Combine("/pathToMyImage/images/", filename); 
            // Noncompliant: Reading data from a file without sanitization.
            return File.ReadAllBytes(filepath);
        }
        // {/fact}
        
        // {fact rule=path-traversal@v1.0 defects=0}
        public static byte[] PathTraversalCompliant(string filename)
        {
            if (string.IsNullOrEmpty(filename)) 
            {   
                throw new ArgumentNullException("error"); 
            }
            filename = Path.GetFileName(filename);
            // Compliant: `Path.GetFileName` used for sanitization.
            string filepath = Path.Combine("/pathToMyImage/images/", filename); 
            return File.ReadAllBytes(filepath);
        }
        // {/fact}
    }
}

